File: 0210-course-schedule-ii/0210-course-schedule-ii.java
class Solution {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        int[] indegree = new int[numCourses];
        int[] order = new int[numCourses];
        ArrayList<ArrayList<Integer>> adj = new ArrayList<>(numCourses);
        
        for(int i =0; i < numCourses; i++){
            adj.add(new ArrayList<>());
        }
        
        for(int[] preq: prerequisites){
            adj.get(preq[1]).add(preq[0]);
            indegree[preq[0]]++;
        }
        
        Deque<Integer> deque = new ArrayDeque<>();
        
        for(int i = 0; i < numCourses; i++){
            if(indegree[i] == 0){
                deque.offer(i);
            }
        }
        
        int nodesVisited = 0;
        int j = 0;
        while(!deque.isEmpty()){
            
            int node = deque.poll();
            nodesVisited++;
            order[j] = node;
            j++;
            
            for(int neighbour: adj.get(node)){
                indegree[neighbour]--;
                if(indegree[neighbour] == 0){
                    deque.offer(neighbour);
                }
            }

            
            indegree[node]--;
            
        }
        int[] empty = {};
        return nodesVisited == numCourses ? order : empty;
    }
}

File: 0210-course-schedule-ii/0210-course-schedule-ii.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        indegree = [0] * numCourses
        order = [0] * numCourses
        adj = [[] for _ in range(numCourses)]

        for preq in prerequisites:
            adj[preq[1]].append(preq[0])
            indegree[preq[0]] += 1

        queue = deque()

        for i in range(numCourses):
            if indegree[i] == 0:
                queue.append(i)

        nodes_visited = 0
        j = 0
        while queue:
            node = queue.popleft()
            nodes_visited += 1
            order[j] = node
            j += 1

            for neighbour in adj[node]:
                indegree[neighbour] -= 1
                if indegree[neighbour] == 0:
                    queue.append(neighbour)

            indegree[node] -= 1

        return order if nodes_visited == numCourses else []